iothub_client/src/iothubtransport_mqtt_common.c

```diff
@@ -1263,357 +1312,434 @@ static int subscribeToNotifyStateIfNeeded(PMQTTTRANSPORT_HANDLE_DATA transport_d
+            if (IoTHubMessage_SetContentTypeSystemProperty(iotHubMessage, propValue) != IOTHUB_MESSAGE_OK)
             {
                 LogError("Failed to set IOTHUB_MESSAGE_HANDLE 'customContentType' property.");
                 result = MU_FAILURE;
             }
-            return result;
+            else
+            {
+                result = 0;
+            }
         }
-        // Codes_SRS_IOTHUB_TRANSPORT_MQTT_COMMON_09_013: [ If type is IOTHUB_TYPE_TELEMETRY and the system property `$.ce` is defined, its value shall be set on the IOTHUB_MESSAGE_HANDLE's ContentEncoding property ]
-        else if (strcmp((const char*)&propName[nameLen - 2], CONTENT_ENCODING_PROPERTY) == 0)
+        break;
+
+        case MQTT_PROPERTY_TYPE_CONTENT_ENCODING:
         {
-            if (IoTHubMessage_SetContentEncodingSystemProperty(IoTHubMessage, propValue) != IOTHUB_MESSAGE_OK)
+            if (IoTHubMessage_SetContentEncodingSystemProperty(iotHubMessage, propValue) != IOTHUB_MESSAGE_OK)
             {
                 LogError("Failed to set IOTHUB_MESSAGE_HANDLE 'contentEncoding' property.");
                 result = MU_FAILURE;
             }
-            return result;
+            else
+            {
+                result = 0;
+            }
+        }
+        break;
+
+        default:
+        {
+            // This is an internal error in code as we should never have an unmapped property at this stage.
+            LogError("Unknown property type = %d to add to message", propertyType);
+            result = MU_FAILURE;
         }
     }
 
     return result;
 }
 
+
 //
-// extractMqttProperties parses the MQTT topic PUBLISH'd to this device/module, retrieves properties and fills out the 
-// IOTHUB_MESSAGE_HANDLE which will ultimately be delivered to the application callback.
+// findMessagePropertyStart takes an MQTT topic PUBLISH'd as input and returns where message properties, if set, begin.
 //
-static int extractMqttProperties(IOTHUB_MESSAGE_HANDLE IoTHubMessage, const char* topic_name, bool urldecode)
+static const char* findMessagePropertyStart(PMQTTTRANSPORT_HANDLE_DATA transportData, const char* topic_name, IOTHUB_IDENTITY_TYPE type)
 {
-    int result;
-    STRING_TOKENIZER_HANDLE token = STRING_TOKENIZER_create_from_char(topic_name);
-    if (token != NULL)
+    const char *propertiesStart;
+
+    if (type == IOTHUB_TYPE_EVENT_QUEUE)
+    {
+        const char* input_queue_topic = STRING_c_str(transportData->topic_InputQueue);
+        // Substract one from queue length to reflect this having an extra # we subscribed to
+        propertiesStart = topic_name + (strlen(input_queue_topic) - 1);
+    }
+    else if (type == IOTHUB_TYPE_TELEMETRY)
+    {
+        const char* mqtt_messsage_queue_topic = STRING_c_str(transportData->topic_MqttMessage);
+        propertiesStart = topic_name + (strlen(mqtt_messsage_queue_topic) - 1);
```

Comment:
might want to add the comment here too about the subtract one for the `#`

Reply:
fixed.